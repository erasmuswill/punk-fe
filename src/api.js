import swr, { useSWRInfinite } from "swr";
import axios from "axios";
import { useContext } from "react";
import { ShowAlcoholicContext } from "./App";
const queryString = require("query-string");

axios.defaults.baseURL = "https://api.punkapi.com/v2";

let cancelSource = axios.CancelToken.source();

const requestInterceptor = (config) => {
  config.cancelToken = cancelSource.token;
  return config;
};

axios.interceptors.request.use(requestInterceptor);

const fetcher = (url) =>
  axios.get(url).then(({ data }) => {
    data = data.filter(
      ({ description, image_url }) => description && image_url
    );
    if (data.length) return data;
    return fetcher(url);
  });
const minorFetcher = (url) =>
  axios.get(url).then(({ data }) => {
    data = data.filter(
      ({ abv, description, image_url }) =>
        abv <= 0.5 && description && image_url
    );
    if (data.length) return data;
    console.log("retrying");
    return minorFetcher(url);
  });

function useSWR(url, isIdempotent) {
  const showAlcoholic = useContext(ShowAlcoholicContext);
  let config = {
    suspense: true,
    loadingTimeout: 1500,
    fetcher: !showAlcoholic && url === "/beers/random" ? minorFetcher : fetcher,
  };
  if (!isIdempotent)
    config = {
      ...config,
      dedupingInterval: 100,
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
    };
  return swr(url, config);
}

function firstIndex(arr) {
  if (arr && Array.isArray(arr) && arr.length) return arr[0];
  else return undefined;
}

function useBeer(id) {
  const { data, error, isValidating, mutate } = useSWR(
    id ? `/beers/${id}` : null
  );

  return {
    beer: data[0],
    isLoading: !error && !data,
    isError: error,
  };
}

function useRandomBeer() {
  const { data, error, isValidating, mutate } = useSWR("/beers/random");

  return {
    beer: firstIndex(data),
    isLoading: !error && !data,
    isError: error,
    isValidating,
    mutate,
  };
}

function useBeerList(filter = []) {
  const showAlcoholic = useContext(ShowAlcoholicContext);
  console.log({ filter });

  const filterObj = {};
  if (filter.length) Object.assign(filterObj, ...filter);

  let { data, error, size, setSize, isValidating } = useSWRInfinite(
    (index) =>
      `/beers?${queryString.stringify({ ...filterObj, page: index + 1 })}`,
    {
      fetcher: (url) =>
        axios.get(url).then(({ data }) => {
          data = data.filter(
            ({ description, image_url }) => description && image_url
          );
          return data;
        }),
      suspense: true,
      loadingTimeout: 1500,
      dedupingInterval: 100,
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
    }
  );
  const isLoadingInitialData = !data && !error,
    isLoadingMore =
      isLoadingInitialData ||
      (size > 0 && data && typeof data[size - 1] === "undefined"),
    hasNextPage = data[data.length - 1].length;

  data = data.filter((v) => v.length).flat();

  if (!showAlcoholic) {
    data = data.filter(({ abv }) => abv <= 0.5);
    if (data.length < 10 && hasNextPage && !isLoadingMore) setSize(size + 1);
  }

  return {
    beers: data,
    isLoadingMore,
    isLoadingInitialData,
    isError: error,
    size,
    setSize,
    isValidating,
    hasNextPage,
  };
}

export { useBeer, useRandomBeer, useBeerList };
