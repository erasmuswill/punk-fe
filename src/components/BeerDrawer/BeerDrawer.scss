.beer-drawer {
  .ant-statistic-title {
    color: #00000065;
  }

  span.ant-descriptions-item-label {
    color: #00000065;
  }

  .ant-descriptions-header {
    margin-bottom: 0px;
    margin-top: 24px;
  }
}
