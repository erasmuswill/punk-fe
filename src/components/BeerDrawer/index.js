import {
  Descriptions,
  Drawer,
  Tooltip,
  Row,
  Col,
  Button,
  Statistic,
  Table,
} from "antd";
import Title from "antd/lib/typography/Title";
import { useState } from "react";
import { useBeer } from "../../api";

function BeerDrawer({ id, close }) {
  const [showIngredients, setShowIngredients] = useState(false);
  const [showMethod, setShowMethod] = useState(false);
  const { beer } = useBeer(id);
  if (!beer) return null;
  const {
    name,
    tagline,
    first_brewed,
    abv,
    ibu,
    target_fg,
    target_og,
    ebc,
    srm,
    ph,
    attenuation_level,
    volume,
    boil_volume,
    method: { mash_temp, twist, fermentation },
    ingredients: { hops, malt, yeast },
    food_pairing,
    brewers_tips,
  } = beer;
  return (
    <Drawer
      title={name}
      width={520}
      onClose={close}
      visible={!!id && id !== 0}
      footer={
        <Row justify="space-around">
          <Col>
            <Button onClick={() => setShowIngredients(true)}>
              Show Ingredients
            </Button>
          </Col>
          <Col>
            <Button onClick={() => setShowMethod(true)}>Show Method</Button>
          </Col>
        </Row>
      }
    >
      <Title level={4}>{tagline}</Title>
      <Row justify="space-between">
        <Col>
          <Statistic title="ABV" value={abv} />
        </Col>
        <Col>
          <Statistic title="IBU" value={ibu} />
        </Col>
        <Col>
          <Statistic title="OG" value={target_og} />
        </Col>
        <Col>
          <Statistic title="First Brewed" value={first_brewed} />
        </Col>
      </Row>
      <Descriptions title="Basics">
        <Descriptions.Item label="Volume">
          <Tooltip title="Amount of wort at the end of the brewing process to be fermented">
            {volume.value} {volume.unit}
          </Tooltip>
        </Descriptions.Item>
        <Descriptions.Item label="Boil Volume">
          <Tooltip
            title="Target amount of liquid to be collected through lautering and sparging
to then be boiled"
          >
            {boil_volume.value} {boil_volume.unit}
          </Tooltip>
        </Descriptions.Item>
        <Descriptions.Item label="ABV">
          <Tooltip title="Alcohol by Volume">{abv}</Tooltip>
        </Descriptions.Item>
        <Descriptions.Item label="Target FG">
          <Tooltip title="Final Gravity; Measurement of relative density of the beer at end of fermentation">
            {target_fg}
          </Tooltip>
        </Descriptions.Item>
        <Descriptions.Item label="Target OG">
          <Tooltip title="Original Gravity; Measurement of the relative density of the wort before fermentation">
            {target_og}
          </Tooltip>
        </Descriptions.Item>
        <Descriptions.Item label="EBC">
          <Tooltip title="Measurement of colour used by european brewing convention">
            {ebc}
          </Tooltip>
        </Descriptions.Item>
        <Descriptions.Item label="SRM">
          <Tooltip title="â€“ Measurement of colour used by american society of brewing chemists">
            {srm}
          </Tooltip>
        </Descriptions.Item>
        <Descriptions.Item label="pH">
          <Tooltip title="Power of Hydrogen; Measurement of acidity or alkalinity">
            {ph}
          </Tooltip>
        </Descriptions.Item>
        <Descriptions.Item label="Attenuation level">
          <Tooltip title="Measurement of the percentage of sugars converted to alcohol and CO2">
            {attenuation_level}
          </Tooltip>
        </Descriptions.Item>
      </Descriptions>
      <Descriptions title="Food Pairings">
        {food_pairing.map((v,i) => (
          <Descriptions.Item key={i}>{v}</Descriptions.Item>
        ))}
      </Descriptions>
      <Descriptions title="Brewer's Tips"></Descriptions>
      <span>{brewers_tips}</span>
      <Drawer
        title="Ingredients"
        width={400}
        onClose={() => setShowIngredients(false)}
        visible={showIngredients}
      >
        <Title level={3}>Malt</Title>
        <Table
          pagination={false}
          dataSource={malt.map(({ amount: { value, unit }, ...rest }) => ({
            amount: `${value} ${unit}`,
            ...rest,
          }))}
          columns={[
            {
              title: "Name",
              dataIndex: "name",
            },
            {
              title: "Amount",
              dataIndex: "amount",
            },
          ]}
        />
        <Title level={3}>Hops</Title>
        <Table
          pagination={false}
          dataSource={hops.map(({ amount: { value, unit }, ...rest }) => ({
            amount: `${value} ${unit}`,
            ...rest,
          }))}
          columns={[
            {
              title: "Name",
              dataIndex: "name",
            },
            {
              title: "Amount",
              dataIndex: "amount",
            },
            {
              title: "Add",
              dataIndex: "add",
            },
            {
              title: "Attribute",
              dataIndex: "attribute",
            },
          ]}
        />
        <Title level={3}>Yeast</Title>
        {yeast}
      </Drawer>
      <Drawer
        title="Method"
        width={400}
        onClose={() => setShowMethod(false)}
        visible={showMethod}
      >
        <Title level={3}>Mash temp</Title>
        <Table
          pagination={false}
          dataSource={mash_temp.map(({ temp: { value, unit }, ...rest }) => ({
            temp: `${value} ${unit}`,
            ...rest,
          }))}
          columns={[
            {
              title: "Temperature",
              dataIndex: "temp",
            },
            {
              title: "Duration",
              dataIndex: "duration",
            },
          ]}
        />
        <Title level={3}>Fermentation</Title>
        <Descriptions>
          <Descriptions.Item label="Temperature">
            {fermentation.temp.value} {fermentation.temp.unit}
          </Descriptions.Item>
        </Descriptions>
        {!!twist && (
          <>
            <Title level={3}>Twist</Title>
            {twist.split(",").map((v, i) => (
              <div key={i}>{v}</div>
            ))}
          </>
        )}
      </Drawer>
    </Drawer>
  );
}
export default BeerDrawer;
