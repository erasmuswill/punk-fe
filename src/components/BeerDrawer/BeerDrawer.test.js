import { fireEvent, render } from "@testing-library/react";
import * as api from "../../api";
import * as antd from "antd";
import BeerDrawer from "./";
import TEST_DATA from "../../TEST_DATA.json";

describe("BeerDrawer", () => {
  it("should not render if no ID", () => {
    const { asFragment } = render(<BeerDrawer />);
    expect(asFragment()).toMatchSnapshot();
  });

  it.skip("should render info", () => {
    jest.spyOn(api, "useBeer").mockImplementation(() => ({
      beer: TEST_DATA.alc_beer,
      loading: false,
      isValidating: false,
    }));
    const { debug } = render(<BeerDrawer id={11} />);
    debug();
    // getByText();
    // expect(asFragment()).toMatchSnapshot();
  });
});
