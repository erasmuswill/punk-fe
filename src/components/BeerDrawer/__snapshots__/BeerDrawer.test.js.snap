// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerDrawer should not render if no ID 1`] = `<DocumentFragment />`;
